pco/src/bit_reader.rs

```diff
@@ -18,50 +17,95 @@ pub unsafe fn u64_at(src: &[u8], byte_idx: usize) -> u64 {
-  //    be useless, so we can read up to (64 - 7) = 57 bits safely from a
-  //    single u64. We right shift by only up to 7 bits, which is safe.
+  // A: We set READ_BYTES so that,
+  //    0  to 25  bit reads -> 4 bytes (1 u32)
+  //    26 to 57  bit reads -> 8 bytes (1 u64)
+  //    58 to 113 bit reads -> 15 bytes (almost 2 u64's)
+  //    For the 1st u64, we read all bytes from the current u64. Due to our bit
+  //    packing, up to the first 7 of these may be useless, so we can read up
+  //    to (64 - 7) = 57 bits safely from a single u64. We right shift by only
+  //    up to 7 bits, which is safe.
   //
   //    For the 2nd u64, we skip only 7 bytes forward. This will overlap with
   //    the 1st u64 by 1 byte, which seems useless, but allows us to avoid one
   //    nasty case: left shifting by U::BITS (a panic). This could happen e.g.
   //    with 64-bit reads when we start out byte-aligned (bits_past_byte=0).
   //
-  //    For the 3rd u64 and onward, we skip 8 bytes forward. Due to how we
-  //    handled the 2nd u64, the most we'll ever need to shift by is
-  //    U::BITS - 8, which is safe.
-  let mut res = U::from_u64(u64_at(src, byte_idx) >> bits_past_byte);
-  let mut processed = min(n, 56 - bits_past_byte);
-  byte_idx += 7;
-
-  for _ in 0..MAX_U64S - 1 {
-    res |= U::from_u64(u64_at(src, byte_idx)) << processed;
-    processed += 64;
-    byte_idx += 8;
+  //    For the 3rd u64 and onward (currently not implemented), we skip 8 bytes
+  //    forward. Due to how we handled the 2nd u64, the most we'll ever need to
+  //    shift by is U::BITS - 8, which is safe.
+
+  match READ_BYTES {
+    4 => read_u32_at(src, byte_idx, bits_past_byte, n),
+    8 => read_u64_at(src, byte_idx, bits_past_byte, n),
+    15 => read_almost_u64x2_at(src, byte_idx, bits_past_byte, n),
+    _ => unreachable!("invalid read bytes: {}", READ_BYTES),
   }
+}
 
-  if MAX_U64S * 64 <= U::BITS as usize {
-    bits::lowest_bits_fast(res, n)
-  } else {
-    bits::lowest_bits(res, n)
-  }
+#[inline]
+unsafe fn read_u32_at<U: ReadWriteUint>(
+  src: &[u8],
+  byte_idx: usize,
+  bits_past_byte: Bitlen,
+  n: Bitlen,
+) -> U {
+  debug_assert!(n <= 25);
+  U::from_u32(bits::lowest_bits_fast(
+    u32_at(src, byte_idx) >> bits_past_byte,
+    n,
+  ))
+}
+
+#[inline]
+unsafe fn read_u64_at<U: ReadWriteUint>(
+  src: &[u8],
+  byte_idx: usize,
+  bits_past_byte: Bitlen,
+  n: Bitlen,
+) -> U {
+  debug_assert!(n <= 57);
+  U::from_u64(bits::lowest_bits_fast(
+    u64_at(src, byte_idx) >> bits_past_byte,
+    n,
+  ))
+}
+
+#[inline]
+unsafe fn read_almost_u64x2_at<U: ReadWriteUint>(
+  src: &[u8],
+  byte_idx: usize,
+  bits_past_byte: Bitlen,
+  n: Bitlen,
+) -> U {
+  debug_assert!(n <= 113);
+  let first_word = u64_at(src, byte_idx) >> bits_past_byte;
+  let processed = 56 - bits_past_byte;
+  let second_word = u64_at(src, byte_idx + 7) << processed;
+  U::from_u64(bits::lowest_bits(
+    first_word | second_word,
```

Comment:
I know it doesn't affect the current set of latent types, but this won't work if we try adding `u128`